diff --git a/src/http/modules/ngx_http_autoindex_module.c b/src/http/modules/ngx_http_autoindex_module.c
index b3bf652..bf3893f 100644
--- a/src/http/modules/ngx_http_autoindex_module.c
+++ b/src/http/modules/ngx_http_autoindex_module.c
@@ -149,6 +149,24 @@ ngx_module_t  ngx_http_autoindex_module = {
 };
 
 
+static size_t ftw_total;
+
+static ngx_int_t
+ftw_add(const char *fpath, const struct stat *sb, int typeflag)
+{
+    ftw_total += sb->st_size;
+    return NGX_OK;
+}
+
+static size_t
+ftw_size(ngx_str_t *path)
+{
+    ftw_total = 0;
+    ftw(path->data, &ftw_add, 512);
+    return ftw_total;
+}
+
+
 static ngx_int_t
 ngx_http_autoindex_handler(ngx_http_request_t *r)
 {
@@ -373,7 +391,7 @@ ngx_http_autoindex_handler(ngx_http_request_t *r)
         entry->dir = ngx_de_is_dir(&dir);
         entry->file = ngx_de_is_file(&dir);
         entry->mtime = ngx_de_mtime(&dir);
-        entry->size = ngx_de_size(&dir);
+        entry->size = ftw_size(&path);
     }
 
     if (ngx_close_dir(&dir) == NGX_ERROR) {
@@ -625,55 +643,44 @@ ngx_http_autoindex_html(ngx_http_request_t *r, ngx_array_t *entries)
                               tm.ngx_tm_min);
 
         if (alcf->exact_size) {
-            if (entry[i].dir) {
-                b->last = ngx_cpymem(b->last,  "                  -",
-                                     sizeof("                  -") - 1);
-            } else {
-                b->last = ngx_sprintf(b->last, "%19O", entry[i].size);
-            }
+            b->last = ngx_sprintf(b->last, "%19O", entry[i].size);
 
         } else {
-            if (entry[i].dir) {
-                b->last = ngx_cpymem(b->last,  "      -",
-                                     sizeof("      -") - 1);
-
-            } else {
-                length = entry[i].size;
-
-                if (length > 1024 * 1024 * 1024 - 1) {
-                    size = (ngx_int_t) (length / (1024 * 1024 * 1024));
-                    if ((length % (1024 * 1024 * 1024))
-                                                > (1024 * 1024 * 1024 / 2 - 1))
-                    {
-                        size++;
-                    }
-                    scale = 'G';
-
-                } else if (length > 1024 * 1024 - 1) {
-                    size = (ngx_int_t) (length / (1024 * 1024));
-                    if ((length % (1024 * 1024)) > (1024 * 1024 / 2 - 1)) {
-                        size++;
-                    }
-                    scale = 'M';
+            length = entry[i].size;
+
+            if (length > 1024 * 1024 * 1024 - 1) {
+                size = (ngx_int_t) (length / (1024 * 1024 * 1024));
+                if ((length % (1024 * 1024 * 1024))
+                                            > (1024 * 1024 * 1024 / 2 - 1))
+                {
+                    size++;
+                }
+                scale = 'G';
 
-                } else if (length > 9999) {
-                    size = (ngx_int_t) (length / 1024);
-                    if (length % 1024 > 511) {
-                        size++;
-                    }
-                    scale = 'K';
+            } else if (length > 1024 * 1024 - 1) {
+                size = (ngx_int_t) (length / (1024 * 1024));
+                if ((length % (1024 * 1024)) > (1024 * 1024 / 2 - 1)) {
+                    size++;
+                }
+                scale = 'M';
 
-                } else {
-                    size = (ngx_int_t) length;
-                    scale = '\0';
+            } else if (length > 9999) {
+                size = (ngx_int_t) (length / 1024);
+                if (length % 1024 > 511) {
+                    size++;
                 }
+                scale = 'K';
 
-                if (scale) {
-                    b->last = ngx_sprintf(b->last, "%6i%c", size, scale);
+            } else {
+                size = (ngx_int_t) length;
+                scale = '\0';
+            }
 
-                } else {
-                    b->last = ngx_sprintf(b->last, " %6i", size);
-                }
+            if (scale) {
+                b->last = ngx_sprintf(b->last, "%6i%c", size, scale);
+
+            } else {
+                b->last = ngx_sprintf(b->last, " %6i", size);
             }
         }
 
@@ -716,9 +723,7 @@ ngx_http_autoindex_json(ngx_http_request_t *r, ngx_array_t *entries,
             + sizeof(", \"type\":\"directory\"") - 1
             + sizeof(", \"mtime\":\"Wed, 31 Dec 1986 10:00:00 GMT\"") - 1;
 
-        if (entry[i].file) {
-            len += sizeof(", \"size\":") - 1 + NGX_OFF_T_LEN;
-        }
+        len += sizeof(", \"size\":") - 1 + NGX_OFF_T_LEN;
     }
 
     b = ngx_create_temp_buf(r->pool, len);
@@ -767,14 +772,9 @@ ngx_http_autoindex_json(ngx_http_request_t *r, ngx_array_t *entries,
 
         b->last = ngx_http_time(b->last, entry[i].mtime);
 
-        if (entry[i].file) {
-            b->last = ngx_cpymem(b->last, "\", \"size\":",
-                                 sizeof("\", \"size\":") - 1);
-            b->last = ngx_sprintf(b->last, "%O", entry[i].size);
-
-        } else {
-            *b->last++ = '"';
-        }
+        b->last = ngx_cpymem(b->last, "\", \"size\":",
+                             sizeof("\", \"size\":") - 1);
+        b->last = ngx_sprintf(b->last, "%O", entry[i].size);
 
         b->last = ngx_cpymem(b->last, " },", sizeof(" },") - 1);
     }
@@ -859,9 +859,7 @@ ngx_http_autoindex_xml(ngx_http_request_t *r, ngx_array_t *entries)
             + entry[i].name.len + entry[i].escape
             + sizeof(" mtime=\"1986-12-31T10:00:00Z\"") - 1;
 
-        if (entry[i].file) {
-            len += sizeof(" size=\"\"") - 1 + NGX_OFF_T_LEN;
-        }
+        len += sizeof(" size=\"\"") - 1 + NGX_OFF_T_LEN;
     }
 
     b = ngx_create_temp_buf(r->pool, len);
@@ -895,11 +893,9 @@ ngx_http_autoindex_xml(ngx_http_request_t *r, ngx_array_t *entries)
                               tm.ngx_tm_mday, tm.ngx_tm_hour,
                               tm.ngx_tm_min, tm.ngx_tm_sec);
 
-        if (entry[i].file) {
-            b->last = ngx_cpymem(b->last, "\" size=\"",
-                                 sizeof("\" size=\"") - 1);
-            b->last = ngx_sprintf(b->last, "%O", entry[i].size);
-        }
+        b->last = ngx_cpymem(b->last, "\" size=\"",
+                             sizeof("\" size=\"") - 1);
+        b->last = ngx_sprintf(b->last, "%O", entry[i].size);
 
         *b->last++ = '"'; *b->last++ = '>';
 
